src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/CompilerTypeSystemContext.IntefaceThunks.cs

```diff
@@ -0,0 +1,287 @@
+        /// </summary>
+        public bool IsDefaultInterfaceMethodImplementationThunkTargetMethod(MethodDesc method)
+        {
+            return method.GetTypicalMethodDefinition().GetType() == typeof(DefaultInterfaceMethodImplementationWithHiddenParameter);
+        }
+
+        /// <summary>
+        /// Returns the real target method of an instantiating thunk.
+        /// </summary>
+        public MethodDesc GetRealDefaultInterfaceMethodImplementationThunkTargetMethod(MethodDesc method)
+        {
+            MethodDesc typicalMethod = method.GetTypicalMethodDefinition();
+            return ((DefaultInterfaceMethodImplementationWithHiddenParameter)typicalMethod).MethodRepresented;
+        }
+
+        private struct DefaultInterfaceMethodImplementationInstantiationThunkHashtableKey
+        {
+            public readonly MethodDesc TargetMethod;
+            public readonly int InterfaceIndex;
+
+            public DefaultInterfaceMethodImplementationInstantiationThunkHashtableKey(MethodDesc targetMethod, int interfaceIndex)
+            {
+                TargetMethod = targetMethod;
+                InterfaceIndex = interfaceIndex;
+            }
+        }
+
+        private class DefaultInterfaceMethodImplementationInstantiationThunkHashtable : LockFreeReaderHashtable<DefaultInterfaceMethodImplementationInstantiationThunkHashtableKey, DefaultInterfaceMethodImplementationInstantiationThunk>
+        {
+            protected override int GetKeyHashCode(DefaultInterfaceMethodImplementationInstantiationThunkHashtableKey key)
+            {
+                return key.TargetMethod.GetHashCode() ^ key.InterfaceIndex;
+            }
+            protected override int GetValueHashCode(DefaultInterfaceMethodImplementationInstantiationThunk value)
+            {
+                return value.TargetMethod.GetHashCode() ^ value.InterfaceIndex;
+            }
+            protected override bool CompareKeyToValue(DefaultInterfaceMethodImplementationInstantiationThunkHashtableKey key, DefaultInterfaceMethodImplementationInstantiationThunk value)
+            {
+                return Object.ReferenceEquals(key.TargetMethod, value.TargetMethod) &&
+                    key.InterfaceIndex == value.InterfaceIndex;
+            }
+            protected override bool CompareValueToValue(DefaultInterfaceMethodImplementationInstantiationThunk value1, DefaultInterfaceMethodImplementationInstantiationThunk value2)
+            {
+                return Object.ReferenceEquals(value1.TargetMethod, value2.TargetMethod) &&
+                    value1.InterfaceIndex == value2.InterfaceIndex;
+            }
+            protected override DefaultInterfaceMethodImplementationInstantiationThunk CreateValueFromKey(DefaultInterfaceMethodImplementationInstantiationThunkHashtableKey key)
+            {
+                TypeDesc owningTypeOfThunks = key.TargetMethod.Context.GeneratedAssembly.GetGlobalModuleType();
```

Comment:
Why not park these on the TargetMethod type?

Reply:
I guess we could do that.

Reply:
Okay, I thought about it a little and it complicates the mangling. Right now we can just make this a Prefix-mangled method (the mangled name of the instantiating thunk is the mangled name of the target default interface method and an integer index into the interface list as part of the prefix). If we make these instance methods on the owning type, the mangled name will need to have the (mangled) owning type, target interface method and the index because there's 3 degrees of freedom. We don't have a nice scheme to produce mangled names for that.

Reply:
Make sense. I forgot about mangling.